### ISSUE TITLE
Missing features of the termination check

### ISSUE THREAD

**Issue by dianegolfouse** [creusot-rs/creusot#1012]
Things missing from #999:
- A way to use structurally decreasing types. E.g.
  ```rust
  enum List<T> {
      Nil,
      Cons(T, Box<List<T>>),
  }
  
  impl<T> List<T> {
      #[terminates]
      fn length(&self) -> usize {
          match self {
              Nil => 0,
              Cons(_, tail) => 1 + tail.length(),
          }
      }
  }
  ```
  
  This can probably be resolved already, by using a correct `#[variant]`. For example for `List<T>`, the variant would be the (logical) length of the list.
- Defer more things to why3, especially for logical functions.

  Since logical functions are fully inlined at their call site, it seems we could just put them all in a big `function f1 ... with f2 ...` block, and let why3 prove that they terminate. But this might be complicated by some functions (like `Vec::deep_model`) having no body.

**Comment by xldenis** [comment]
> Since logical functions are fully inlined at their call site, it seems we could just put them all in a big function f1 ... with f2 ... block, and let why3 prove that they terminate.

There's a corollary issue: we don't want to duplicate proofs, and termination proofs are indissociable from the rest of the correctness proof for a logical function. The bodies we clone for logical functions do not generate proof obligations (which is on purpose), if a logical function has a variant we will not generate its body in the module which makes use of it, instead we will provide a defining axiom which accounts for the variant. 

Something which could potentially help here is if we could tell why3 to assume the termination of a function (since we have externally proved it). This sounds wildly unsound though.

**Comment by xldenis** [comment]
> By default, we were not tracking termination for program functions.

Right! I feel like `variants` have _sort_ of worked for program functions in the past though, but I'm much less confident there.
